skill_predictor.py

```diff
@@ -0,0 +1,210 @@
+                            tmp_prev_skills = itertools.permutations(
+                                list(prev_skills))
+                            for permutation in tmp_prev_skills:
+                                curr_sentence = list(permutation) + [skill]
+                                curr_sentence = ' '.join(curr_sentence)
+                                if np.random.rand() < 0.8:
+                                    f.write('%s\n' % curr_sentence)
+                                else:
+                                    g.write('%s\n' % curr_sentence)
+                                counter += 1
+                                if counter > num_permutations:
+                                    break
+
+                # skill deletions
+                for skill in prev_skills:
+                    if skill not in curr_skills:
+                        with open('train_data.txt', 'a') as f, open('test_data.txt', 'a') as g:
+                            counter = 0
+                            tmp_prev_skills = itertools.permutations(
+                                list(prev_skills))
+                            for permutation in tmp_prev_skills:
+                                curr_sentence = list(
+                                    permutation) + ['-' + skill]
+                                curr_sentence = ' '.join(curr_sentence)
+                                if np.random.rand() < 0.8:
+                                    f.write('%s\n' % curr_sentence)
+                                else:
+                                    g.write('%s\n' % curr_sentence)
+                                counter += 1
+                                if counter > num_permutations:
+                                    break
+
+    def train(self):
+        """
+        Trains the given KenLM model on a language model dataset
+        """
+        try:
+            self.model = kenlm.Model('../kenlm/build/train_data.arpa')
+        except:
+            print('Data file not found')
+            print(
+                'Run generate_data() and move the train_data.txt to ../kenlm/build/ folder')
+            print('In ../kenlm/build/ folder run the following command in Terminal: bin/lmplz -o 6 < train_data.txt > train_data.arpa --discount_fallback')
+            print('Try again')
+
+    def predict(self, sentence):
+        """
+        Predicts the next bag of skills given a current bag of skills as a sentence
+        """
+        if self.model is None:
+            print('Run train() first to train the model')
+            return
+        heap = []
+        next_bag = []
+
+        # add deletion skills
+        prediction_skills = list(self.all_skills)[:]
+        negation_skills = ['-' + skill for skill in self.all_skills]
+        prediction_skills.extend(negation_skills)
+
+        skills_list = sentence.split()
+        prev_bag_len = len(skills_list)
+        if prev_bag_len in self.mle_bagsize:
+            counter = max(
```

Comment:
Why max? Wouldn't it be bestter to sample from the distribution? Otherwise we are not generating the right varianc on the bag lengths...